### PR TITLE
gh-102856: Add changes related to PEP 701 in 3.12 What's New docs

### PR DESCRIPTION
**Pull request by mgmacias95**
* Issue: gh-102856

----
:books: Documentation preview :books:: https://cpython-previews--104824.org.readthedocs.build/

### CODE REVIEW

**Comment by pablogsal** [comment]
Let's make sure we cover also the changes being mentioned in https://github.com/python/cpython/issues/104802

**Comment by pablogsal** [comment]
@CAM-Gerlach could you help us copy-editing the text if you have sone free cycles? :)

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -136,22 +136,70 @@ Improved Error Messages
...
+unicode escaped sequences, multi-line expressions, comments and strings reusing the
+same quote as the containing f-string. Let's cover these in detail:
+
+* Quote reuse: in Python 3.11, reusing the same quotes as the contaning f-string
+  raises a :exc:`SyntaxError`, forcing the user to either use other available
+  quotes (like using double quotes or triple quites if the f-string uses single
+  quites). In Python 3.12, you can now do things like this:
+
+    >>> songs = ['Take me back to Eden', 'Alkaline', 'Ascensionism']
+    >>> f"This is the playlist: {", ".join(things)}"
```
Comment (pablogsal) [inline Doc/whatsnew/3.12.rst]:
```suggestion
    >>> f"This is the playlist: {", ".join(songs)}"
```

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -136,22 +136,70 @@ Improved Error Messages
...
+
+PEP 701: Syntactic formalization of f-strings
+---------------------------------------------
+
+:pep:`701` lifts some restrictions on the usage of f-strings. Expression components
+inside f-strings can now be any valid Python expression including backslashes,
+unicode escaped sequences, multi-line expressions, comments and strings reusing the
+same quote as the containing f-string. Let's cover these in detail:
+
+* Quote reuse: in Python 3.11, reusing the same quotes as the contaning f-string
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* Quote reuse: in Python 3.11, reusing the same quotes as the containing f-string
```

File: Doc/whatsnew/3.12.rst [hunk 3]
```diff
@@ -136,22 +136,70 @@ Improved Error Messages
...
+  raises a :exc:`SyntaxError`, forcing the user to either use other available
+  quotes (like using double quotes or triple quites if the f-string uses single
+  quites). In Python 3.12, you can now do things like this:
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  quotes (like using double quotes or triple quotes if the f-string uses single
  quotes). In Python 3.12, you can now do things like this:
```

File: Doc/whatsnew/3.12.rst [hunk 4]
```diff
@@ -136,22 +136,70 @@ Improved Error Messages
...
+
+    >>> songs = ['Take me back to Eden', 'Alkaline', 'Ascensionism']
+    >>> f"This is the playlist: {", ".join(songs)}"
+    'This is the playlist: Take me back to Eden, Alkaline, Ascensionism
+
+  Note that before this change there was no explicit limit in how f-strings can
+  be nested, but the fact that string quotes cannot be reused inside the
+  expression component of f-strings made it impossible to nest f-strings
+  arbitrarily. In fact, this is the most nested-fstring that can be written:
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  arbitrarily. In fact, this is the most nested f-string that could be written:
```

File: Doc/whatsnew/3.12.rst [hunk 5]
```diff
@@ -1201,6 +1278,34 @@ Changes in the Python API
   that may be surprising or dangerous.
   See :ref:`tarfile-extraction-filter` for details.
 
+* The output of the :func:`tokenize.tokenize` and :func:`tokenize.generate_tokens`
+  functions is now changed due to the changes introduced in :pep:`701`. This
+  means that ``STRING`` tokens are not emited anymore for f-strings and the
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  means that ``STRING`` tokens are not emitted any more for f-strings and the
```

File: Doc/whatsnew/3.12.rst [hunk 6]
```diff
@@ -1201,6 +1278,34 @@ Changes in the Python API
...
+  tokens described in :pep:`701` are now produced instead: ``FSTRING_START``,
+  ``FSRING_MIDDLE`` and ``FSTRING_END`` are now emited for f-string "string"
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  ``FSRING_MIDDLE`` and ``FSTRING_END`` are now emitted for f-string "string"
```

File: Doc/whatsnew/3.12.rst [hunk 7]
```diff
@@ -1201,6 +1278,34 @@ Changes in the Python API
...
+  parts in addition to the the apropiate tokens for the tokenization in the
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  parts in addition to the appropriate tokens for the tokenization in the
```

File: Doc/whatsnew/3.12.rst [hunk 8]
```diff
@@ -1201,6 +1278,34 @@ Changes in the Python API
...
+      1,2-1,8:            FSTRING_MIDDLE 'start '
+      1,8-1,9:            OP             '{'
+      1,9-1,10:           NUMBER         '1'
+      1,10-1,11:          OP             '+'
+      1,11-1,12:          NUMBER         '1'
+      1,12-1,13:          OP             '}'
+      1,13-1,17:          FSTRING_MIDDLE ' end'
+      1,17-1,18:          FSTRING_END    '"'
+
+  Aditionally, final ``DEDENT`` tokens are now emited within the bounds of the
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  Aditionally, final ``DEDENT`` tokens are now emitted within the bounds of the
```

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
Mergin now gives visibility to this change to users looking at the What's New for 3.12 beta 1

**Comment by hugovk** [comment]
This broke the CI's doctest check, here's a PR to fix it: https://github.com/python/cpython/pull/104852

File: Doc/whatsnew/3.12.rst [hunk 9]
```diff
@@ -136,22 +136,70 @@ Improved Error Messages
...
+inside f-strings can now be any valid Python expression including backslashes,
+unicode escaped sequences, multi-line expressions, comments and strings reusing the
+same quote as the containing f-string. Let's cover these in detail:
+
+* Quote reuse: in Python 3.11, reusing the same quotes as the contaning f-string
+  raises a :exc:`SyntaxError`, forcing the user to either use other available
+  quotes (like using double quotes or triple quites if the f-string uses single
+  quites). In Python 3.12, you can now do things like this:
+
+    >>> songs = ['Take me back to Eden', 'Alkaline', 'Ascensionism']
```
Comment (brentropic) [inline Doc/whatsnew/3.12.rst:158]: Very good improvement.

**Pull request merged**